### PR TITLE
Implement transaction dao

### PR DESCRIPTION
**Pull request by seregamazur**

### CODE REVIEW

File: src/main/java/com/smartcity/dao/TransactionDaoImpl.java [hunk 1]
```diff
@@ -1,22 +1,88 @@
...
+        try {
+            template.update(Queries.SQL_TRANSACTION_CREATE,
+                    transaction.getTaskId(),
+                    transaction.getCurrentBudget(),
+                    transaction.getTransactionBudget(),
+                    CurrentDate.getCurDate(),
+                    transaction.getUpdatedDate());
+            return Optional.of(transaction);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/TransactionDaoImpl.java]: Remove this line. User logging instead

File: src/main/java/com/smartcity/dao/TransactionDaoImpl.java [hunk 2]
```diff
@@ -1,22 +1,88 @@
...
+    private interface Queries {
+        String SQL_TRANSACTION_CREATE = "INSERT INTO Transactions(task_id,current_budget,transaction_budget," +
+                "created_date,updated_date) values(?,?,?,?,?)";
+        String SQL_TRANSACTION_GET = "SELECT * FROM Transactions where id = ?";
+        String SQL_TRANSACTION_UPDATE = "UPDATE Transactions set task_id = ? , current_budget = ? ," +
+                "transaction_budget = ? ,created_date = ? , updated_date = ? where id = ?";
+        String SQL_TRANSACTION_DELETE = "DELETE FROM Transactions where id = ?";
+    }
+
+    private static class CurrentDate {
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/TransactionDaoImpl.java]: Use java.date.LocalDateTime instead

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.datasource.DriverManagerDataSource;
+
+import java.sql.Date;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class TransactionDaoImplTest {
```
Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: make the class public

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        dataSource.setUrl("jdbc:mysql://192.168.99.100:3306/smartcity");
+        dataSource.setUsername("root");
+        dataSource.setPassword("root");
+        addInsertions();
+        transDao = new TransactionDaoImpl(dataSource);
+    }
+
+
+    @Test
+    void create() {
```
Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: Name test methods like: testCrateTransaction, testUpdateTransaction, testUpdateTransaction_invalidId, testCreateTransaction_missingTaskId, etc...

Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: make all test methods public

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Test
+    void delete() {
+        transDao.create(transaction);
+        transDao.delete(1);
+        assertThrows(RuntimeException.class, () -> transDao.get(1).get());
+
+    }
+
+    private void addInsertions() {
```
Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: In a future move it to DaoTestBase class that @Igor-Polatajko added

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    @Test
+    public void testGetTransaction() {
+        transDao.create(transaction);
+        assertThat(transaction).isEqualToIgnoringGivenFields(transDao.get(1L).get(),
+                "createdDate", "updatedDate");
+    }
+
+    @Test
+    public void testGetTransaction_invalidId() {
+        transDao.create(transaction);
```
Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: assert this method returns true

File: src/main/java/com/smartcity/dao/TransactionDaoImpl.java [hunk 7]
```diff
@@ -1,22 +1,85 @@
...
+    private JdbcTemplate template;
+
+    @Autowired
+    public TransactionDaoImpl(DataSource source) {
+        template = new JdbcTemplate(source);
+    }
 
-    public Transaction create(Transaction transaction) {
-        return null;
+    public Optional<Transaction> create(Transaction transaction) {
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/dao/TransactionDaoImpl.java]: Implement entity creation similar to that one:
https://github.com/iholovatyi/smartcity/pull/3/files#diff-10dd1bb1c445f3fb05099b12729442ceR31
with autogenerated id

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        assertThat(transDao.get(1).get()).isEqualToIgnoringGivenFields(transaction,
+                "createdDate", "updatedDate");
+        assertThrows(RuntimeException.class, () -> transDao.get(Long.MAX_VALUE));
+    }
+
+    @Test
+    public void testDeleteTransaction() {
+        transDao.create(transaction);
+        transDao.delete(1);
+        assertThrows(RuntimeException.class, () -> transDao.get(1).get());
```
Comment (seregamazur) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: Assert false

File: src/main/java/com/smartcity/mapper/TransactionMapper.java [hunk 9]
```diff
@@ -2,12 +2,30 @@
...
+    private TransactionMapper() {
+    }
+
+    public static TransactionMapper getInstance() {
+        return instance;
+    }
+
     public Transaction mapRow(ResultSet resultSet, int i) throws SQLException {
-        return null;
+        Transaction transaction = new Transaction();
```
Comment (iholovatyi) [inline src/main/java/com/smartcity/mapper/TransactionMapper.java]: Return null if resultSet is null

File: src/test/java/com/smartcity/dao/TransactionDaoImplTest.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        transDao.update(newTransaction);
+        assertThat(transDao.get(1).get()).isEqualToIgnoringGivenFields(transaction,
+                "createdDate", "updatedDate");
+        assertThrows(RuntimeException.class, () -> transDao.get(Long.MAX_VALUE));
+    }
+
+    @Test
+    public void testDeleteTransaction() {
+        transDao.create(transaction);
+        transDao.delete(1);
```
Comment (iholovatyi) [inline src/test/java/com/smartcity/dao/TransactionDaoImplTest.java]: assert true

**Pull request merged**